File: bindings/julia/src/Httpd.jl
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# Julia bindings for the proven-httpd protocol (HTTP server).
#
# Wraps the C-ABI functions from protocols/proven-httpd/ffi/zig/src/httpd.zig
# via ccall into libproven_httpd.so.

module Httpd

using ..ProvenServers: check_status, check_slot, SlotId

export HttpMethod, RequestPhase, HttpVersion, StatusCode, ParseResult,
       abi_version, create_context, destroy_context, parse_request,
       get_method, get_phase, get_version, set_status, send_response,
       keep_alive_check, reset_context, can_transition

const LIB = "libproven_httpd"

# --------------------------------------------------------------------------
# Enumeration types matching Idris2 ABI
# --------------------------------------------------------------------------

"""HTTP methods matching the Zig ABI tag values."""
@enum HttpMethod::UInt8 begin
    METHOD_GET     = 0
    METHOD_HEAD    = 1
    METHOD_POST    = 2
    METHOD_PUT     = 3
    METHOD_DELETE  = 4
    METHOD_CONNECT = 5
    METHOD_OPTIONS = 6
    METHOD_TRACE   = 7
    METHOD_PATCH   = 8
end

"""HTTP request lifecycle phases matching the Zig ABI."""
@enum RequestPhase::UInt8 begin
    PHASE_IDLE           = 0
    PHASE_RECEIVING      = 1
    PHASE_HEADERS_PARSED = 2
    PHASE_BODY_RECEIVING = 3
    PHASE_COMPLETE       = 4
    PHASE_RESPONDING     = 5
    PHASE_SENT           = 6
end

"""HTTP protocol versions."""
@enum HttpVersion::UInt8 begin
    HTTP_1_0 = 0
    HTTP_1_1 = 1
end

"""HTTP status code tags."""
@enum StatusCode::UInt8 begin
    STATUS_200_OK           = 0
    STATUS_201_CREATED      = 1
    STATUS_204_NO_CONTENT   = 2
    STATUS_301_MOVED        = 3
    STATUS_302_FOUND        = 4
    STATUS_304_NOT_MODIFIED = 5
    STATUS_400_BAD_REQUEST  = 6
    STATUS_401_UNAUTHORIZED = 7
    STATUS_403_FORBIDDEN    = 8
    STATUS_404_NOT_FOUND    = 9
    STATUS_405_NOT_ALLOWED  = 10
    STATUS_500_INTERNAL     = 11
    STATUS_502_BAD_GATEWAY  = 12
    STATUS_503_UNAVAILABLE  = 13
end

"""Parse result codes."""
@enum ParseResult::UInt8 begin
    PARSE_COMPLETE  = 0
    PARSE_REJECTED  = 1
    PARSE_NEED_MORE = 2
end

# --------------------------------------------------------------------------
# ccall declarations
# --------------------------------------------------------------------------

"""Return the ABI version of the linked libproven_httpd."""
function abi_version()::UInt32
    ccall((:http_abi_version, LIB), UInt32, ())
end

"""
    create_context() -> SlotId

Create a new HTTP context in the Idle phase. Throws on pool exhaustion.
"""
function create_context()::SlotId
    raw = ccall((:http_create_context, LIB), Cint, ())
    check_slot(raw)
end

"""
    destroy_context(slot::SlotId)

Release the given HTTP context slot.
"""
function destroy_context(slot::SlotId)::Nothing
    ccall((:http_destroy_context, LIB), Cvoid, (Cint,), slot)
    nothing
end

"""
    parse_request(slot::SlotId, data::Vector{UInt8}) -> ParseResult

Feed raw HTTP data into a context for parsing.
"""
function parse_request(slot::SlotId, data::Vector{UInt8})::ParseResult
    raw = ccall((:http_parse_request, LIB), UInt8,
                (Cint, Ptr{UInt8}, UInt32),
                slot, data, UInt32(length(data)))
    ParseResult(raw)
end

"""
    get_method(slot::SlotId) -> Union{HttpMethod, Nothing}

Get the HTTP method of the parsed request. Returns nothing if not yet parsed.
"""
function get_method(slot::SlotId)::Union{HttpMethod, Nothing}
    tag = ccall((:http_get_method, LIB), UInt8, (Cint,), slot)
    tag == 0xff ? nothing : HttpMethod(tag)
end

"""
    get_phase(slot::SlotId) -> Union{RequestPhase, Nothing}

Get the current request processing phase.
"""
function get_phase(slot::SlotId)::Union{RequestPhase, Nothing}
    tag = ccall((:http_get_phase, LIB), UInt8, (Cint,), slot)
    tag > UInt8(PHASE_SENT) ? nothing : RequestPhase(tag)
end

"""
    get_version(slot::SlotId) -> Union{HttpVersion, Nothing}

Get the HTTP version of the parsed request.
"""
function get_version(slot::SlotId)::Union{HttpVersion, Nothing}
    tag = ccall((:http_get_version, LIB), UInt8, (Cint,), slot)
    tag > UInt8(HTTP_1_1) ? nothing : HttpVersion(tag)
end

"""
    set_status(slot::SlotId, status::StatusCode)

Set the response status code. Throws on invalid state.
"""
function set_status(slot::SlotId, status::StatusCode)::Nothing
    raw = ccall((:http_set_status, LIB), UInt8,
                (Cint, UInt8), slot, UInt8(status))
    check_status(raw)
end

"""
    send_response(slot::SlotId)

Send the response, transitioning Responding -> Sent. Throws on invalid state.
"""
function send_response(slot::SlotId)::Nothing
    raw = ccall((:http_send_response, LIB), UInt8, (Cint,), slot)
    check_status(raw)
end

"""
    keep_alive_check(slot::SlotId) -> Bool

Check if the connection uses keep-alive.
"""
function keep_alive_check(slot::SlotId)::Bool
    ccall((:http_keep_alive_check, LIB), UInt8, (Cint,), slot) == 0x01
end

"""
    reset_context(slot::SlotId)

Reset the context for keep-alive reuse (Sent -> Idle). Throws on invalid state.
"""
function reset_context(slot::SlotId)::Nothing
    raw = ccall((:http_reset_context, LIB), UInt8, (Cint,), slot)
    check_status(raw)
end

"""
    can_transition(from::RequestPhase, to::RequestPhase) -> Bool

Stateless query: check whether a lifecycle transition is valid.
"""
function can_transition(from::RequestPhase, to::RequestPhase)::Bool
    ccall((:http_can_transition, LIB), UInt8,
          (UInt8, UInt8), UInt8(from), UInt8(to)) == 0x01
end

end # module Httpd

File: bindings/javascript/src/httpd.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// JavaScript bindings for the proven-httpd Zig FFI.
//
// Wraps the C-ABI functions from protocols/proven-httpd/ffi/zig/src/httpd.zig:
//   - Context lifecycle: http_create_context, http_destroy_context
//   - Request parsing: http_parse_request
//   - Request queries: http_get_method, http_get_path, http_get_header, http_get_body
//   - Response construction: http_set_status, http_set_header, http_set_body,
//     http_send_response
//   - Phase & transition: http_get_phase, http_get_version, http_keep_alive_check,
//     http_reset_context, http_can_transition

import { checkSlot, checkStatus } from "./error.js";
import { loadLibrary } from "./ffi.js";

// ---------------------------------------------------------------------------
// Enums matching Idris2 ABI tags
// ---------------------------------------------------------------------------

/** @readonly @enum {number} HTTP request methods. */
export const Method = Object.freeze({
    GET: 0, POST: 1, PUT: 2, DELETE: 3, PATCH: 4,
    HEAD: 5, OPTIONS: 6, TRACE: 7, CONNECT: 8,
});

/** @readonly @enum {number} HTTP request lifecycle phases. */
export const RequestPhase = Object.freeze({
    IDLE: 0, RECEIVING: 1, HEADERS_PARSED: 2, BODY_RECEIVING: 3,
    COMPLETE: 4, RESPONDING: 5, SENT: 6,
});

/** @readonly @enum {number} HTTP response status code tags. */
export const StatusCode = Object.freeze({
    OK: 0, CREATED: 1, NO_CONTENT: 2, MOVED_PERMANENTLY: 3, FOUND: 4,
    NOT_MODIFIED: 5, BAD_REQUEST: 6, UNAUTHORIZED: 7, FORBIDDEN: 8,
    NOT_FOUND: 9, METHOD_NOT_ALLOWED: 10, CONFLICT: 11, GONE: 12,
    UNPROCESSABLE_ENTITY: 13, TOO_MANY_REQUESTS: 14,
    INTERNAL_SERVER_ERROR: 15, NOT_IMPLEMENTED: 16, BAD_GATEWAY: 17,
    SERVICE_UNAVAILABLE: 18, GATEWAY_TIMEOUT: 19,
});

/** @readonly @enum {number} HTTP version tags. */
export const Version = Object.freeze({
    HTTP_1_0: 0, HTTP_1_1: 1, HTTP_2: 2,
});

/** @readonly @enum {number} Parse result values. */
export const ParseResult = Object.freeze({
    COMPLETE: 0, REJECTED: 1, NEED_MORE: 2,
});

// ---------------------------------------------------------------------------
// Module state
// ---------------------------------------------------------------------------

/** @type {object|null} */
let _lib = null;

/**
 * Initialize the httpd FFI library.
 *
 * @param {object} [options] - Loading options passed to loadLibrary.
 * @returns {Promise<void>}
 */
export async function init(options) {
    if (!_lib) {
        _lib = await loadLibrary("httpd", options);
    }
}

/** @returns {object} The loaded library, throwing if not initialized. */
function lib() {
    if (!_lib) throw new Error("httpd: call init() before using the module");
    return _lib;
}

// ---------------------------------------------------------------------------
// Context class
// ---------------------------------------------------------------------------

/**
 * HTTP request/response context wrapping a Zig FFI slot.
 *
 * The slot is released when destroy() is called. Use the static create()
 * factory for async initialization.
 *
 * @example
 * await init();
 * const ctx = HttpdContext.create();
 * try {
 *   const result = ctx.parseRequest(rawData);
 *   if (result === ParseResult.COMPLETE) {
 *     ctx.setStatus(StatusCode.OK);
 *     ctx.setBody(new TextEncoder().encode("Hello"));
 *     ctx.sendResponse();
 *   }
 * } finally {
 *   ctx.destroy();
 * }
 */
export class HttpdContext {
    /** @param {number} slot */
    constructor(slot) {
        /** @type {number} */
        this._slot = slot;
        /** @type {boolean} */
        this._destroyed = false;
    }

    /**
     * Create a new HTTP context.
     * @returns {HttpdContext}
     */
    static create() {
        const slot = checkSlot(lib().http_create_context());
        return new HttpdContext(slot);
    }

    /** Release the context slot back to the pool. */
    destroy() {
        if (!this._destroyed) {
            lib().http_destroy_context(this._slot);
            this._destroyed = true;
        }
    }

    /**
     * Feed raw HTTP data into the context for parsing.
     * @param {Uint8Array} data - Raw HTTP request bytes.
     * @returns {number} ParseResult value.
     */
    parseRequest(data) {
        return lib().http_parse_request(this._slot, data, data.length);
    }

    /**
     * Get the HTTP method of the parsed request.
     * @returns {number|null} Method tag, or null if not parsed (255).
     */
    getMethod() {
        const tag = lib().http_get_method(this._slot);
        return tag === 255 ? null : tag;
    }

    /**
     * Get the request path.
     * @param {number} [maxLen=4096] - Maximum path length.
     * @returns {string} The request path.
     */
    getPath(maxLen = 4096) {
        const buf = new Uint8Array(maxLen);
        const written = lib().http_get_path(this._slot, buf, maxLen);
        return new TextDecoder().decode(buf.subarray(0, written));
    }

    /**
     * Look up a request header by key (case-insensitive).
     * @param {string} key - Header name.
     * @param {number} [maxLen=4096] - Maximum value length.
     * @returns {string} Header value, or empty string if not found.
     */
    getHeader(key, maxLen = 4096) {
        const keyBuf = new TextEncoder().encode(key);
        const valBuf = new Uint8Array(maxLen);
        const written = lib().http_get_header(
            this._slot, keyBuf, keyBuf.length, valBuf, maxLen,
        );
        return new TextDecoder().decode(valBuf.subarray(0, written));
    }

    /**
     * Get the request body.
     * @param {number} [maxLen=65536] - Maximum body length.
     * @returns {Uint8Array} The request body bytes.
     */
    getBody(maxLen = 65536) {
        const buf = new Uint8Array(maxLen);
        const written = lib().http_get_body(this._slot, buf, maxLen);
        return buf.subarray(0, written);
    }

    /**
     * Set the response status code.
     * @param {number} status - StatusCode tag.
     */
    setStatus(status) {
        checkStatus(lib().http_set_status(this._slot, status));
    }

    /**
     * Set a response header.
     * @param {string} key - Header name.
     * @param {string} value - Header value.
     */
    setHeader(key, value) {
        const k = new TextEncoder().encode(key);
        const v = new TextEncoder().encode(value);
        checkStatus(lib().http_set_header(this._slot, k, k.length, v, v.length));
    }

    /**
     * Set the response body.
     * @param {Uint8Array} data - Response body bytes.
     */
    setBody(data) {
        checkStatus(lib().http_set_body(this._slot, data, data.length));
    }

    /** Send the response. Transitions Responding -> Sent. */
    sendResponse() {
        checkStatus(lib().http_send_response(this._slot));
    }

    /**
     * Check if the connection uses keep-alive.
     * @returns {boolean}
     */
    keepAliveCheck() {
        return lib().http_keep_alive_check(this._slot) === 1;
    }

    /**
     * Get the current request processing phase.
     * @returns {number|null} RequestPhase tag, or null if unknown.
     */
    getPhase() {
        const tag = lib().http_get_phase(this._slot);
        return tag <= 6 ? tag : null;
    }

    /**
     * Get the HTTP version.
     * @returns {number|null} Version tag, or null if unknown.
     */
    getVersion() {
        const tag = lib().http_get_version(this._slot);
        return tag <= 2 ? tag : null;
    }

    /** Reset context for keep-alive reuse (Sent -> Idle). */
    reset() {
        checkStatus(lib().http_reset_context(this._slot));
    }
}

// ---------------------------------------------------------------------------
// Module-level functions
// ---------------------------------------------------------------------------

/**
 * Return the ABI version of the linked libproven_httpd.
 * @returns {number}
 */
export function abiVersion() {
    return lib().http_abi_version();
}

/**
 * Stateless query: check whether a lifecycle transition is valid.
 * @param {number} from - Source RequestPhase tag.
 * @param {number} to - Target RequestPhase tag.
 * @returns {boolean}
 */
export function canTransition(from, to) {
    return lib().http_can_transition(from, to) === 1;
}
